### PR TITLE
RSDK-3345 - getPointcloudMap refactor

### PR DESCRIPTION
**Pull request by kim-mishra**

### CODE REVIEW

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 1]
```diff
@@ -51,6 +52,14 @@ int CalculateProbabilityFromColorChannels(ColorARGB pixel_color) {
...
 namespace viam {
 
+std::string std_string_from_bstring(bstring b_str) {
+    int len = blength(b_str);
+    char *tmp = bstr2cstr(b_str, 0);
+    std::string std_str(tmp, len);
+    bcstrfree(tmp);
+    return std_str;
+}
+
```
Comment (kim-mishra) [inline viam-cartographer/src/slam_service/slam_service.cc:62]: TODO: import from `carto_facade.h` getting compile errors as of rn

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 2]
```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:203]: We can't call `std::terminate()` as doing so will terminate the entire program. We need to return an informative error code.

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 3]
```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+                LOG(INFO)
+                    << "In localization mode, using cached pointcloud map";
+            } else {
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:200]: We need to know the expected & unexpected exceptions that can be received here, & classify them into return codes.

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 4]
```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return 1;
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:208]: please use constants, not raw integers.

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 5]
```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+    }
+
+    response->point_cloud_pcd =
+        blk2bstr((const char *)pointcloud_map.c_str(), pointcloud_map.length());
+
+    return 0;
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:214]: please use constants, not raw integers.

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 6]
```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return 1;
+    }
+
+    response->point_cloud_pcd =
+        blk2bstr((const char *)pointcloud_map.c_str(), pointcloud_map.length());
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:212]: Is this cast really necessary? You don't seem to need to do it here: https://github.com/viamrobotics/viam-cartographer/pull/121/files#diff-92ff8a5cd3a57430d78c6ba3b61a91d1810399b77613eece3c55841868d2822eR214

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 7]
```diff
@@ -89,47 +98,20 @@ ::grpc::Status SLAMServiceImpl::GetPosition(ServerContext *context,
...
-        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
-                   << e.what();
-        std::terminate();
-    }
+    viam_carto_get_point_cloud_map_response vcgpcmr;
+    int status = GetPointCloudMapC(&vcgpcmr);
 
-    if (pointcloud_map.empty()) {
-        LOG(ERROR) << "map pointcloud does not have points yet";
+    if (status == 1) {
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:104]: please use constants

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 8]
```diff
@@ -186,6 +178,56 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+            } else {
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::runtime_error &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
```
Comment (kim-mishra) [inline viam-cartographer/src/slam_service/slam_service.cc:212]: If we don't want to catch all `runtime_errors`, we can change the type of exception thrown just thought we would want to change as little as possible for now, but let me know if you think we should be throwing more specific errors

Comment (nicksanford) [reply viam-cartographer/src/slam_service/slam_service.cc:212]: If we needed to handle different runtime errors with different return codes we could look into creating a class per runtime error, however what you have here is appropriate for our current use case.

**Review by nicksanford — COMMENTED** [review]
This looks good to me. Holding off on approval until this is moved to be on CartoFacade.

File: viam-cartographer/src/slam_service/slam_service.cc (hunk 8 above)
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:212]:
```suggestion
        LOG(ERROR) << e.what();
```

File: viam-cartographer/src/slam_service/slam_service.cc [hunk 9]
```diff
@@ -186,6 +178,56 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
...
+        return VIAM_CARTO_POINTCLOUD_ENCODING_ERROR;
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
```
Comment (nicksanford) [inline viam-cartographer/src/slam_service/slam_service.cc:216]:
```suggestion
        LOG(ERROR) << e.what();
```

**Review by nicksanford — CHANGES_REQUESTED** [review]
Actually just a few requests to remove language from logs that describe shutting down. We are no longer shutting down when we hit these conditions.

**Pull request closed without merging**